python/tvm/relax/transform/op_legalizer.py

```diff
@@ -0,0 +1,561 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=abstract-method,invalid-name,missing-module-docstring
+import logging
+
+import tvm
+from tvm import ir, te, topi, relax
+from tvm.relax import struct_info
+from tvm.ir.module import IRModule
+
+from ..analysis import remove_all_unused
+from ..expr import Call, Function, ShapeExpr, Tuple, TupleGetItem, Var
+from ..expr_functor import mutator, PyExprMutator
+from ..block_builder import BlockBuilder
+
+
+# Todo(ruihang): confirm symbolic shape support
+
+
+def has_known_shape_value(sinfo: struct_info.StructInfo):
+    if isinstance(sinfo, struct_info.TensorStructInfo):
```

Comment:
document a bit

Reply:
Yes it is in progress. I’d better change it to a draft PR.